src/main/java/com/google/cloud/pubsublite/spark/PslWriteDataSourceOptions.java

```diff
@@ -0,0 +1,136 @@
+ */
+
+package com.google.cloud.pubsublite.spark;
+
+import static com.google.cloud.pubsublite.internal.ExtractStatus.toCanonical;
+import static com.google.cloud.pubsublite.internal.wire.ServiceClients.addDefaultMetadata;
+import static com.google.cloud.pubsublite.internal.wire.ServiceClients.addDefaultSettings;
+
+import com.google.api.gax.rpc.ApiException;
+import com.google.auto.value.AutoValue;
+import com.google.cloud.pubsublite.AdminClient;
+import com.google.cloud.pubsublite.AdminClientSettings;
+import com.google.cloud.pubsublite.MessageMetadata;
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.TopicPath;
+import com.google.cloud.pubsublite.internal.Publisher;
+import com.google.cloud.pubsublite.internal.wire.PartitionCountWatchingPublisherSettings;
+import com.google.cloud.pubsublite.internal.wire.PubsubContext;
+import com.google.cloud.pubsublite.internal.wire.RoutingMetadata;
+import com.google.cloud.pubsublite.internal.wire.SinglePartitionPublisherBuilder;
+import com.google.cloud.pubsublite.spark.internal.PslCredentialsProvider;
+import com.google.cloud.pubsublite.v1.AdminServiceClient;
+import com.google.cloud.pubsublite.v1.AdminServiceSettings;
+import com.google.cloud.pubsublite.v1.PublisherServiceClient;
+import com.google.cloud.pubsublite.v1.PublisherServiceSettings;
+import java.io.Serializable;
+import javax.annotation.Nullable;
+import org.apache.spark.sql.sources.v2.DataSourceOptions;
+
+@AutoValue
+public abstract class PslWriteDataSourceOptions implements Serializable {
+
+  @Nullable
+  public abstract String credentialsKey();
+
+  public abstract TopicPath topicPath();
+
+  public static Builder builder() {
+    return new AutoValue_PslWriteDataSourceOptions.Builder().setCredentialsKey(null);
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract PslWriteDataSourceOptions.Builder setCredentialsKey(String credentialsKey);
+
+    public abstract PslWriteDataSourceOptions.Builder setTopicPath(TopicPath topicPath);
+
+    public abstract PslWriteDataSourceOptions build();
+  }
+
+  public static PslWriteDataSourceOptions fromSparkDataSourceOptions(DataSourceOptions options) {
+    if (!options.get(Constants.TOPIC_CONFIG_KEY).isPresent()) {
+      throw new IllegalArgumentException(Constants.TOPIC_CONFIG_KEY + " is required.");
+    }
+
+    Builder builder = builder();
+    String topicPathVal = options.get(Constants.TOPIC_CONFIG_KEY).get();
+    try {
+      builder.setTopicPath(TopicPath.parse(topicPathVal));
+    } catch (ApiException e) {
+      throw new IllegalArgumentException("Unable to parse topic path " + topicPathVal, e);
+    }
+    options.get(Constants.CREDENTIALS_KEY_CONFIG_KEY).ifPresent(builder::setCredentialsKey);
+    return builder.build();
+  }
+
+  public PslCredentialsProvider getCredentialProvider() {
+    return new PslCredentialsProvider(credentialsKey());
+  }
+
+  public static Publisher<MessageMetadata> createNewPublisher(
```

Comment:
There's no reason for a static method with this object as its first argument. Just make it a normal method. Ditto the below.